numpyro/hmc_util.py

```diff
@@ -268,3 +271,321 @@ def update_fn(t, accept_prob, z_flat, state):
+# ref: https://github.com/google/jax/issues/446
+# TODO: remove this when namedtuple is supported in JAX
+register_pytree_node(
+    _TreeInfo,
+    lambda xs: (tuple(xs), None),
+    lambda _, xs: _TreeInfo(*xs)
+)
+
+
+def _is_turning(inverse_mass_matrix, r_left, r_right, r_sum):
+    r_left, _ = ravel_pytree(r_left)
+    r_right, _ = ravel_pytree(r_right)
+    r_sum, _ = ravel_pytree(r_sum)
+
+    if inverse_mass_matrix.ndim == 2:
+        v_left = np.matmul(inverse_mass_matrix, r_left)
+        v_right = np.matmul(inverse_mass_matrix, r_right)
+    elif inverse_mass_matrix.ndim == 1:
+        v_left = np.multiply(inverse_mass_matrix, r_left)
+        v_right = np.multiply(inverse_mass_matrix, r_right)
+
+    # This implements dynamic termination criterion (ref [2], section A.4.2).
+    turning_at_left = np.dot(v_left, r_sum - r_left) <= 0
+    turning_at_right = np.dot(v_right, r_sum - r_right) <= 0
+    return turning_at_left | turning_at_right
+
+
+def _uniform_transition_prob(current_subtree, new_subtree, rng):
+    # This function computes transition prob for subtrees (ref [2], section A.3.1).
+    # e^new_weight / (e^new_weight + e^current_weight)
+    transition_prob = expit(new_subtree.weight - current_subtree.weight)
+    return transition_prob
+
+
+def _biased_transition_prob(current_tree, new_tree, rng):
+    # This function computes transition prob for main trees (ref [2], section A.3.2).
+    transition_prob = np.exp(new_tree.weight - current_tree.weight)
+    # If new tree is turning or diverging, we won't move the proposal
+    # to the new tree.
+    transition_prob = np.where(new_tree.turning | new_tree.diverging,
+                               0.0, np.clip(transition_prob, a_max=1.0))
+    return transition_prob
+
+
+@partial(jit, static_argnums=(6,))
+def _combine_tree(current_tree, new_tree, inverse_mass_matrix, going_right, rng,
+                  biased_transition, iterative_build):
+    # Now we combine the current tree and the new tree. Note that outside
+    # leaves of the combined tree are determined by the direction.
+    z_left, r_left, z_left_grad, z_right, r_right, r_right_grad = lax.cond(
+        going_right,
+        (current_tree, new_tree),
+        lambda trees: (trees[0].z_left, trees[0].r_left,
+                       trees[0].z_left_grad, trees[1].z_right,
+                       trees[1].r_right, trees[1].z_right_grad),
+        (new_tree, current_tree),
+        lambda trees: (trees[0].z_left, trees[0].r_left,
+                       trees[0].z_left_grad, trees[1].z_right,
+                       trees[1].r_right, trees[1].z_right_grad)
+    )
+
+    transition_prob = lax.cond(biased_transition,
+                               (current_tree, new_tree, rng),
+                               lambda args: _biased_transition_prob(*args),
+                               (current_tree, new_tree, rng),
+                               lambda args: _uniform_transition_prob(*args))
```

Comment:
@neerajprad  I make `biased_transition` becomes a dynamic variable instead of a static fn `_biased_transition_prob`/`_uniform_transition_prob`. This saves a bit of time after compiling.